casper/src/main/scala/coop/rchain/casper/api/BlockApiImpl.scala

```diff
@@ -600,10 +600,16 @@ class BlockApiImpl[F[_]: Concurrent: RuntimeManager: BlockDagStorage: BlockStore
       par: Par,
       blockHash: String,
       usePreStateHash: Boolean
-  ): F[ApiErr[(Seq[Par], LightBlockInfo)]] =
-    for {
-      block     <- BlockStore[F].getUnsafe(blockHash.unsafeHexToByteString)
-      sortedPar <- parSortable.sortMatch[F](par).map(_.term)
-      data      <- getDataWithBlockInfo(sortedPar, block).map(_.get)
-    } yield (data.postBlockData, data.block).asRight[Error]
+  ): F[ApiErr[(Seq[Par], LightBlockInfo)]] = {
+    val loadData = (for {
+      block     <- OptionT(BlockStore[F].get1(blockHash.unsafeHexToByteString))
+      sortedPar <- OptionT.liftF(parSortable.sortMatch[F](par).map(_.term))
+      stateHash = if (usePreStateHash) block.body.state.preStateHash
+      else block.body.state.postStateHash
+      data <- OptionT.liftF(RuntimeManager[F].getData(stateHash)(sortedPar))
+      lbi  = getLightBlockInfo(block)
+    } yield (data, lbi))
+    if (blockHash.isEmpty) "Invalid block hash".asLeft[(Seq[Par], LightBlockInfo)].pure[F]
+    else loadData.value.map(_.toRight[Error](s"No block with $blockHash."))
```

Comment:
You can just raise exceptions for each wrong data and catch them at the end to convert to Left. Now that conversion to Either and string errors is done here (on the wrong place too early). 
```scala
val data = for {
  a <- aOpt.liftTo(new Exception(...))
  b <- bOpt.liftTo(new Exception(...))
  c <- cOpt.liftTo(new Exception(...))
} yield ...
data.map(_.toRight).handleErrorWith { case ex => ex.toLeft }
```